pyccl/parameters.py

```diff
@@ -0,0 +1,121 @@
+from . import ccllib as lib
+from .errors import warnings, CCLDeprecationWarning
+
+
+class CCLParameters:
+    """Base for classes holding global CCL parameters and their values.
+
+    Subclasses contain a reference to the C-struct with the collection
+    of parameters and their values (via SWIG). All subclasses act as proxies
+    to the CCL parameters at the C-level.
+
+    Subclasses automatically store a backup of the initial parameter state
+    to enable ad hoc reloading.
+    """
+
+    def __init_subclass__(cls, instance=None, freeze=False):
+        """Routine for subclass initialization.
+
+        Parameters:
+            instance (``pyccl.ccllib.cvar``):
+                Reference to ``cvar`` where the parameters are implemented.
+            freeze (``bool``):
+                Disable parameter mutation.
+        """
+        super().__init_subclass__()
+        cls._instance = instance
+        cls._frozen = freeze
+
+        def _new_setattr(self, key, value):
+            # Make instances of the SWIG-level class immutable
+            # so that everything is handled through this interface.
+            # SWIG only assigns `this` via the low level `_ccllib`;
+            # we therefore disable all other direct assignments.
+            if key == "this":
+                return object.__setattr__(self, key, value)
+            name = self.__class__.__name__
+            # TODO: Deprecation cycle for fully immutable Cosmology objects.
+            # raise AttributeError(f"Direct assignment in {name} not supported.")  # noqa
+            warnings.warn(
+                f"Direct assignment of {name} is deprecated "
+                "and an error will be raised in the next CCL release."
+                f"Set via `pyccl.{name}.{key}` before instantiation.",
+                CCLDeprecationWarning)
+            object.__setattr__(self, key, value)
+
+        cls._instance.__class__.__setattr__ = _new_setattr
+
+    def __init__(self):
+        # Keep a copy of the default parameters.
+        object.__setattr__(self, "_bak", CCLParameters.get_params_dict(self))
+
+    def __getattribute__(self, name):
+        get = object.__getattribute__
+        try:
+            return get(get(self, "_instance"), name)
+        except AttributeError:
+            return get(self, name)
+
+    def __setattr__(self, key, value):
+        if self._frozen and key != "T_CMB":
+            # `T_CMB` mutates in Cosmology.
+            name = self.__class__.__name__
+            raise AttributeError(f"Instances of {name} are frozen.")
+        if not hasattr(self._instance, key):
+            raise KeyError(f"Parameter {key} does not exist.")
+        if (key, value) == ("A_SPLINE_MAX", 1.0):
```

Comment:
`A_SPLINE_MAX` should be immutable.

Reply:
It is. Here, for example I am making use of the C-code. If `A_SLINE_MAX` gets set to 1, then the function returns doing nothing. If not, the C-code takes over and raises.

Reply:
Could you make it raise an exception even when being set to 1? `A_MAX_SPLINE` shouldn't be touched, even when trying to assign the "correct" value.

Reply:
I would argue it's okay to set it to its correct value. It is actually touched (set) whenever the parameters are reloaded to their default values, through `spline_params.reload()`.

Also in the (near) future, we might want to relax this requirement to see how things evolve at scale factors larger than 1 (i.e. the far future).

Reply:
Jokes aside, if we have it raise an error, I would have to fall back to `object.__setattr__` in `reload`, which is slightly slower.

Reply:
Fine. I'd just caution that the C code implicitly assumes `A_MAX_SPLINE==1` in a couple of places and allowing setting `A_MAX_SPLINE` needs a careful review of much of the C code base to make sure this is okay.